File: src/simulation/demography.js
import { recordEvent } from "./journal.js";
import { recalculateSpatialAggregates } from "./grid-spatial.js";

function setCityPopulation(world, cityId, targetPopulation) {
  const node = world.spatial.nodes[`city:${cityId}`];
  const territories = node.childTerritoryIds.map((id) => world.spatial.territories[id]);
  const current = territories.reduce((sum, territory) => sum + territory.population, 0);
  const weights = territories.map((territory) => Math.max(0.1, territory.population +
    (territory.id === node.anchorTerritoryId ? Math.max(10, current * 0.01) : 0)));
  const totalWeight = weights.reduce((sum, weight) => sum + weight, 0);
  let assigned = 0;
  const fractions = [];
  for (let index = 0; index < territories.length; index += 1) {
    const exact = targetPopulation * weights[index] / totalWeight;
    territories[index].population = Math.floor(exact);
    assigned += territories[index].population;
    fractions.push({ territory: territories[index], fraction: exact - Math.floor(exact) });
  }
  fractions.sort((left, right) => right.fraction - left.fraction ||
    left.territory.id.localeCompare(right.territory.id));
  for (let index = 0; index < targetPopulation - assigned; index += 1) {
    fractions[index].territory.population += 1;
  }
}

function monthlyVitalChange(world, city) {
  const policy = world.demographyPolicy;
  const population = world.spatial.nodes[city.spatialNodeId].aggregate.population;
  const birthExpected = population * policy.birthRatePerYear * 30 / world.calendar.daysPerYear +
    city.demography.birthRemainder;
  const mortalityFactor = (1.22 - city.demography.health * 0.28) *
    (city.shortage.active ? policy.shortageMortalityMultiplier : 1) *
    (Object.values(city.needs).some((need) => need.active) ? 1.18 : 1);
  const deathExpected = population * policy.deathRatePerYear * mortalityFactor * 30 /
    world.calendar.daysPerYear + city.demography.deathRemainder;
  const births = Math.floor(birthExpected);
  const deaths = Math.min(population + births, Math.floor(deathExpected));
  city.demography.birthRemainder = birthExpected - births;
  city.demography.deathRemainder = deathExpected - deaths;
  city.demography.births += births;
  city.demography.deaths += deaths;
  const infrastructureHealth = (city.infrastructure.housingCondition + city.infrastructure.sanitation) / 2;
  city.demography.health = Math.max(0.1, Math.min(1,
    city.demography.health + (infrastructureHealth - city.demography.health) * 0.035 +
      (city.shortage.active ? -0.025 : Object.values(city.needs).some((need) => need.active) ? -0.006 : 0.002)
  ));
  return { population, births, deaths, target: population + births - deaths };
}

function migrationDestination(world, sourceId) {
  const neighbors = world.routes
    .filter((route) => route.a === sourceId || route.b === sourceId)
    .map((route) => route.a === sourceId ? route.b : route.a)
    .filter((cityId) => !world.cities[cityId].shortage.active);
  return neighbors.sort((leftId, rightId) => {
    const left = world.cities[leftId];
    const right = world.cities[rightId];
    const leftPopulation = world.spatial.nodes[left.spatialNodeId].aggregate.population;
    const rightPopulation = world.spatial.nodes[right.spatialNodeId].aggregate.population;
    const leftCoverage = left.stocks.food / Math.max(0.001, leftPopulation * left.foodPerPersonPerDay);
    const rightCoverage = right.stocks.food / Math.max(0.001, rightPopulation * right.foodPerPersonPerDay);
    return rightCoverage - leftCoverage || leftId.localeCompare(rightId);
  })[0] ?? null;
}

export function advanceDemography(world) {
  if (world.day === 0 || world.day % 30 !== 0) return;
  const changes = new Map();
  for (const cityId of Object.keys(world.cities).sort()) {
    changes.set(cityId, monthlyVitalChange(world, world.cities[cityId]));
  }

  for (const cityId of Object.keys(world.cities).sort()) {
    const city = world.cities[cityId];
    if (!city.shortage.active || city.shortage.episodeDays < 10) continue;
    const destinationId = migrationDestination(world, cityId);
    if (!destinationId) continue;
    const change = changes.get(cityId);
    const migrants = Math.min(change.target, Math.max(1,
      Math.floor(change.population * world.demographyPolicy.monthlyMigrationShare)
    ));
    change.target -= migrants;
    changes.get(destinationId).target += migrants;
    city.demography.emigration += migrants;
    world.cities[destinationId].demography.immigration += migrants;
    recordEvent(world, {
      type: "migration_flow",
      subjectId: cityId,
      causeIds: city.shortage.eventId ? [city.shortage.eventId] : [],
      details: { from: cityId, to: destinationId, people: migrants }
    });
  }

  for (const [cityId, change] of changes) setCityPopulation(world, cityId, change.target);
  recalculateSpatialAggregates(world);

  if (world.day % world.calendar.daysPerYear < 30) {
    for (const [cityId, change] of changes) {
      recordEvent(world, {
        type: "population_report",
        subjectId: cityId,
        details: {
          cityId,
          population: change.target,
          birthsToDate: world.cities[cityId].demography.births,
          deathsToDate: world.cities[cityId].demography.deaths
        }
      });
    }
  }
}

File: src/WorldGen.Core/Simulation/Demography.cs
using System.Text.Json.Nodes;

namespace WorldGen.Core.Simulation;

public static class Demography
{
    public static void Advance(WorldState world)
    {
        if (world.Day == 0 || world.Day % 30 != 0) return;
        var changes = world.Cities.Keys.Order(StringComparer.Ordinal)
            .ToDictionary(id => id, id => MonthlyChange(world, world.Cities[id]), StringComparer.Ordinal);
        foreach (var cityId in world.Cities.Keys.Order(StringComparer.Ordinal))
        {
            var city = world.Cities[cityId];
            if (!city.Shortage.Active || city.Shortage.EpisodeDays < 10) continue;
            var destination = MigrationDestination(world, cityId); if (destination is null) continue;
            var change = changes[cityId];
            var migrants = Math.Min(change.Target, Math.Max(1, (int)Math.Floor(change.Population * world.DemographyPolicy.MonthlyMigrationShare)));
            change.Target -= migrants; changes[destination].Target += migrants;
            city.Demography.Emigration += migrants; world.Cities[destination].Demography.Immigration += migrants;
            Journal.Record(world, "migration_flow", cityId, city.Shortage.EventId is null ? [] : [city.Shortage.EventId],
                new JsonObject { ["from"] = cityId, ["to"] = destination, ["people"] = migrants });
        }
        foreach (var pair in changes) SetCityPopulation(world, pair.Key, pair.Value.Target);
        SpatialRuntime.RecalculateAggregates(world);
        if (world.Day % world.Calendar.DaysPerYear < 30)
        foreach (var pair in changes)
            Journal.Record(world, "population_report", pair.Key, details:
                new JsonObject { ["cityId"] = pair.Key, ["population"] = pair.Value.Target,
                    ["birthsToDate"] = world.Cities[pair.Key].Demography.Births, ["deathsToDate"] = world.Cities[pair.Key].Demography.Deaths });
    }

    private static VitalChange MonthlyChange(WorldState world, CityState city)
    {
        var policy = world.DemographyPolicy; var population = world.Spatial.Nodes[city.SpatialNodeId].Aggregate.Population;
        var birthExpected = population * policy.BirthRatePerYear * 30 / world.Calendar.DaysPerYear + city.Demography.BirthRemainder;
        var mortality = (1.22 - city.Demography.Health * 0.28) * (city.Shortage.Active ? policy.ShortageMortalityMultiplier : 1) *
            (city.Needs.Values.Any(need => need.Active) ? 1.18 : 1);
        var deathExpected = population * policy.DeathRatePerYear * mortality * 30 / world.Calendar.DaysPerYear + city.Demography.DeathRemainder;
        var births = (int)Math.Floor(birthExpected); var deaths = Math.Min(population + births, (int)Math.Floor(deathExpected));
        city.Demography.BirthRemainder = birthExpected - births; city.Demography.DeathRemainder = deathExpected - deaths;
        city.Demography.Births += births; city.Demography.Deaths += deaths;
        var infrastructureHealth = (city.Infrastructure.HousingCondition + city.Infrastructure.Sanitation) / 2;
        city.Demography.Health = SimulationMath.Clamp(city.Demography.Health + (infrastructureHealth - city.Demography.Health) * 0.035 +
            (city.Shortage.Active ? -0.025 : city.Needs.Values.Any(need => need.Active) ? -0.006 : 0.002), 0.1, 1);
        return new VitalChange(population, population + births - deaths);
    }

    private static string? MigrationDestination(WorldState world, string sourceId) => world.Routes
        .Where(route => route.A == sourceId || route.B == sourceId).Select(route => route.A == sourceId ? route.B : route.A)
        .Where(id => !world.Cities[id].Shortage.Active)
        .OrderByDescending(id =>
        {
            var city = world.Cities[id]; var population = world.Spatial.Nodes[city.SpatialNodeId].Aggregate.Population;
            return city.Stocks["food"] / Math.Max(0.001, population * city.FoodPerPersonPerDay);
        }).ThenBy(id => id, StringComparer.Ordinal).FirstOrDefault();

    private static void SetCityPopulation(WorldState world, string cityId, int target)
    {
        var node = world.Spatial.Nodes[$"city:{cityId}"]; var territories = node.ChildTerritoryIds!.Select(id => world.Spatial.Territories[id]).ToArray();
        var current = territories.Sum(item => item.Population);
        var weights = territories.Select(item => Math.Max(0.1, item.Population + (item.Id == node.AnchorTerritoryId ? Math.Max(10, current * 0.01) : 0))).ToArray();
        var totalWeight = weights.Sum(); var assigned = 0; var fractions = new List<(Spatial.Territory Territory, double Fraction)>();
        for (var i = 0; i < territories.Length; i++)
        {
            var exact = target * weights[i] / totalWeight; territories[i].Population = (int)Math.Floor(exact);
            assigned += territories[i].Population; fractions.Add((territories[i], exact - Math.Floor(exact)));
        }
        fractions.Sort((a, b) => { var c = b.Fraction.CompareTo(a.Fraction); return c != 0 ? c : SimulationMath.LocaleComparer.Compare(a.Territory.Id, b.Territory.Id); });
        for (var i = 0; i < target - assigned; i++) fractions[i].Territory.Population++;
    }

    private sealed record VitalChange(int Population, int InitialTarget) { public int Target { get; set; } = InitialTarget; }
}
